Personal recommendations page

sigmike opened the pull request:
Related issue: #363

I started the personal recommendations page, mostly copied from the "compared items" page. It can be accessed by URL only for now, at `/personal_recommendations/<username>`.

Since the API only returns entity uids I had to fetch each entity individually (whereas the "compared items" API returns the entities directly). It's not a problem but it may be slower for the user.

sigmike commented on frontend/scripts/openapi.yaml:
```diff
@@ -1973,6 +1973,7 @@ components:
       - tournesol_score
       - type
       - uid
+      - metadata
```
I added that as required field because otherwise an `EntityCard` doesn't accept an `Entity` object because `metadata` may be undefined. But I'm not sure the backend always includes this attributes. If it doesn't we can also change `EntityCard` to make it handle an undefined `metadata`.

GresilleSiffle replied:
This file is generated by `yarn update-schema` so if you've manually added `metadata` I'm scared it will be automatically removed the next time the command is ran.

I'll probably be working on something else until the end of the week, so I can't suggest a proper solution for now, but let's keep in mind this should be fixed before merging.

sigmike replied:
If we change the API it won't be necessary anymore because the new API can return an `EntityNoExtraField`. But it's probably cleaner anyway to make the common attributes between `EntityNoExtraField` and `Entity` have the same requiredness.

sigmike commented:
I removed the personal recommendations page and added a checkbox on the ratings page, as suggested in https://github.com/tournesol-app/tournesol/issues/363#issuecomment-1082349950:

![image](https://user-images.githubusercontent.com/28259/160871086-71cd7e6c-c3df-4273-94ed-ad1e3d0283ad.png)

GresilleSiffle commented:
Hélo @sigmike 

I'm adding a comment now to not forget about it before the review / suggestion.

> Since the API only returns entity uids I had to fetch each entity individually (whereas the "compared items" API returns the entities directly). It's not a problem but it may be slower for the user.

That might be a problem indeed, as the back end doesn't allow an unlimited number of HTTP requests per user per minute and per user per hour to slow down potential attacks. If the endpoint doesn't return what you need in the front end we should consider updating the back end accordingly if relevant.

I don't remember if you are familiar with Python and Django, but if not don't worry we will take care of it :+1:

sigmike commented:
I'm not familiar with Django but I'm interested to learn. I will try to do it.

sigmike commented:
I changed the API to return the entities directly.

The serializer is based on the `Entity` model, so I just added the extra fields from `EntityNoExtraField` instead of adding an `entity` attribute (that would make it similar to the ratings API). It may be cleaner to make the serializer more similar to the ratings API, but it looks like a big change.

I made all the serializer fields read only to make them `required` in `openapi.yaml` (otherwise the `VideoCard` complains about `metadata` being potentially undefined). I don't know if there's a better way, nor what this "read only" attribute means exactly, but it doesn't look inappropriate for this API anyway.

Having an `entity` attribute instead of the fields directly in the result would also allow to remove `ContributorRecommendations` from the `RelatedEntityObject` type. It probably doesn't hurt having it there but it looks weird.

GresilleSiffle commented:
Thanks for taking the time to update the back end as well.

The current back end implementation looks a bit complex to me, and may make the contributor recommendation serializer more difficult to maintain now.

Before this PR the `ContributorRecommendationsSerializer` inherited from `RecommendationSerializer`. Now it still inherits from `RecommendationSerializer` but also include an extra field from another source `EntityNoExtraFieldSerializer`. This is a bit hacky for several reasons:

**(1)** it won't be obvious for a developer that updating `EntityNoExtraFieldSerializer` will also update `ContributorRecommendationsSerializer` but not `RecommendationSerializer`

**(2)** it's weird that `ContributorRecommendationsSerializer` requires field from a serializer that should not contain extra fields (`EntityNoExtraFieldSerializer`)

**(3)** inheriting fields from two different sources in the meta class of `ContributorRecommendationsSerializer` adds a complexity that can be avoided.

Inheriting from `EntityNoExtraFieldSerializer` adds the `type` field in the response. I'm going to see if this field is required to display a contributor's recommendations, and if yes it should be added directly to the `ContributorRecommendationsSerializer` meta class without inheritance, or inherited from `RecommendationSerializer` is relevant.

GresilleSiffle commented on backend/tournesol/tests/test_api_contributor_recommendations.py:
```diff
@@ -333,3 +334,18 @@ def test_recommendations_are_sorted_by_descending_total_score(self):
                 response.data["results"][i]["uid"], score_to_entity[score].uid
             )
+
+    def test_recommendations_include_entity_no_extra_field_fields(self):
```
This test may become obsolete if we remove the extra field added in `ContributorRecommendationSerializer`

GresilleSiffle replied:
After some tests, you can directly check the presence of the `type` field in the test  `test_api_polls.PollsRecommendationsApi.test_anonymous_can_list_recommendations` , and remove this test.

With something like this:

```python
self.assertEqual(results[0]["type"], "video")
```

GresilleSiffle commented on backend/tournesol/serializers/contributor_recommendations.py:
```diff
@@ -20,7 +21,9 @@ class Meta(RecommendationSerializer.Meta):
             - {"n_comparisons"}
             - {"n_contributors"}
             | {"is_public"}
+            | set(EntityNoExtraFieldSerializer.Meta.fields)
```
Alright, I've checked the front end is this field is not required to display anything, but only to make the response of the contributor's recommendations API compliant with the type `RelatedEntityObject`.

Suggestion:

1. add the field `type` directly in the `RecommendationSerializer`
2. this will make all recommendations API compliant with the front end type `RelatedEntityObject`
3. delete the `| ContributorRecommendations` in the front end file `src/utils/types.ts`

and everything should be OK on the back end side : )

sigmike replied:
The idea behind using `EntityNoExtraFieldSerializer` was that if new fields are added to it, and required by `VideoCard`, then we would not have to change this serializer too (and the same could be applied to all serializers that produce something that must be sent to a `VideoCard`). Kind of defining an interface for serializers whose result is used in video cards. Using `EntityNoExtraFieldSerializer` for that is probably not the best way though. I understand it adds complexity, but duplicating the required fields on each serializer doesn't seem ideal either.

I think it would be cleaner to have an extra `entity` field that would use a standard serializer that always works with video cards.

Anyway I applied your suggestions in 2db1bc40d957018f15333c5576cadef9b69c3bd1. I still had to make `metadata` read only to make it required.

I fail to understand why we can now remove `ContributorRecommendations` from `RelatedEntityObject` though. If it's because it now has all the same fields than `EntityNoExtraField` and `RelatedEntity` so typescript doesn't care, then I don't understand why I had to add it before (I also made it have all the same fields).

GresilleSiffle replied:
> I understand it adds complexity, but duplicating the required fields on each serializer doesn't seem ideal either.

I think we are good, as the complexity added by the original proposition is far more important than configuring two serializers with the same required fields. This way they remain uncoupled.

> I think it would be cleaner to have an extra entity field that would use a standard serializer that always works with video cards.

Hmm, I'm not sure I understand, maybe we could talk about it quickly on Friday after the meeting?

> I fail to understand why we can now remove ContributorRecommendations from RelatedEntityObject though. If it's because it now has all the same fields than EntityNoExtraField and RelatedEntity so typescript doesn't care, then I don't understand why I had to add it before (I also made it have all the same fields).

You're right, you don't need `ContributorRecommendations` anymore in the front end because, by adding `type`, you made the entities returned by the recommendations API (both global and personal) compliant with what the front end is  expecting.

On `main` the page `page/videos/VideoRatings` only works with the rating API which has one subtle difference with the recommendations APIs: the presence of the field `type`. This difference was not intended, it's just we never have needed the field `type` until now.

The back end should as much as possible returns the same representation for an entity, regardless of the endpoint. These difference with `type` could be seen as a design bug. Maybe the `RecommendationsSerializer` should inherit its "base" fields from the `EntitySerializer`, and maybe all endpoints dealing with entities should be based on `EntitySerializer` to avoid having the same error in the future.

sigmike replied:
> This way they remain uncoupled.

I agree my proposition was not good but I still think all the serializers whose output is used in video cards *should* be coupled somehow. If we change the video card to display a new entity attribute for example, it would be better not to have to change multiple serializers to add this field.

> > I think it would be cleaner to have an extra entity field that would use a standard serializer that always works with video cards.
> Hmm, I'm not sure I understand, maybe we could talk about it quickly on Friday after the meeting?

Sure.

My suggestion is that the API would work like `users_me_contributor_ratings_list`: it would return something with an `entity` field, instead of returning an entity with additional fields. That way we could use a generic serializer to produce the `entity` field (like `ContributorRatingSerializer` uses `EntityNoExtraFieldSerializer` to serialize the entity).

> The back end should as much as possible returns the same representation for an entity, regardless of the endpoint. 

Indeed, that's the reason for my points above.

> Maybe the RecommendationsSerializer should inherit its "base" fields from the EntitySerializer, and maybe all endpoints dealing with entities should be based on EntitySerializer to avoid having the same error in the future.

Yes that would work too to avoid having to change multiple serializers when we want to change how an entity is serialized. I think having the entities always serialized the same way (or a few ways) would be slightly cleaner though (compared to having multiple API returning entities with some common fields and some different extra fields). But I'm not sure which one is preferable in practice.

sigmike replied:
These changes can be discussed or done in a separate issue or PR.

GresilleSiffle replied:
> These changes can be discussed or done in a separate issue or PR.

Yes sure, it's really out of the initial scope : )

GresilleSiffle reviewed (dismissed):
Hello @sigmike

I hit the request changes button because I saw we could make the back end implementation more simple.

GresilleSiffle commented:
Currently reading the front end part...

GresilleSiffle reviewed (commented):
LGTM, and it works :+1: 

Thanks @sigmike . Sorry I haven't been available the last Friday. I'm just going to check few details with Adrien and then I will merge.
